File: simplex.java/src/main/java/kr/co/tmon/simplex/store/Store.java
package kr.co.tmon.simplex.store;


import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

import java.lang.reflect.Field;
import java.util.Collection;
import java.util.Hashtable;
import java.util.concurrent.TimeUnit;

import kr.co.tmon.simplex.Simplex;
import kr.co.tmon.simplex.actions.IActionBinder;
import kr.co.tmon.simplex.actions.IActionBinderSet;
import kr.co.tmon.simplex.annotations.Unsubscribe;
import kr.co.tmon.simplex.channels.Channel;
import kr.co.tmon.simplex.channels.IChannel;
import kr.co.tmon.simplex.reactivex.Unit;
import kr.co.tmon.simplex.utils.ChannelUtil;
import kr.co.tmon.simplex.utils.CloneUtil;
import kr.co.tmon.simplex.actions.IAction;
import kr.co.tmon.simplex.actions.TransformedResult;
import kr.co.tmon.simplex.actions.descriptor.IDispatchDescriptor;
import kr.co.tmon.simplex.actions.descriptor.ISubscriptionDescriptor;

public class Store<TActionBinderSet extends IActionBinderSet> implements IActionStore<TActionBinderSet> {

	private Class<TActionBinderSet> clazz;
	
	private TActionBinderSet actions;
	
	private Hashtable<String, Object> observableSources;
	
	public Store(Class<TActionBinderSet> clazz) {
		this.clazz = clazz;
		this.observableSources = new Hashtable<String, Object>();
	}
	
	protected TActionBinderSet getActions() throws InstantiationException, IllegalAccessException {
		if (actions == null) {
			actions = clazz.newInstance();
		}

		return actions;
	}
	
	@Override
	public <TAction extends IAction<TParam, TResult>, TParam, TResult> void dispatch(
			Function<DescriptorFactory<TActionBinderSet>, IDispatchDescriptor<TAction, TParam, TResult>> descriptor) {
		try {
			DispatchDescriptor<TAction, TParam, TResult> desc 
				= (DispatchDescriptor<TAction, TParam, TResult>)descriptor.apply(new DescriptorFactory<TActionBinderSet>(getActions()));
			
			dispatchInner(
					desc.getAction(), 
					desc.getParameter(), 
					desc.getBegin(), 
					desc.getEnd(), 
					desc.getChannel());
			
		} catch (Exception e) {
			Simplex.ExceptionSubject.onNext(e);
			Simplex.Logger.write(e);
		}
	}
	
	
	@Override
	public <TAction extends IAction<TParam, TResult>, TParam, TResult> Disposable subscribe(
			Function<DescriptorFactory<TActionBinderSet>, ISubscriptionDescriptor<TAction, TParam, TResult>> descriptor) { 
		
		Disposable disposable = null;
		try {
			SubscriptionDescriptor<TAction, TParam, TResult> desc 
				= (SubscriptionDescriptor<TAction, TParam, TResult>)descriptor.apply(new DescriptorFactory<TActionBinderSet>(getActions()));
			
			disposable = subscribeInner(
					desc.getAction(), 
					desc.getOnNext(), 
					desc.getObserveOnMainThread(),
					desc.getObservable(),
					desc.getChannel(),
					desc.getPreventClone());
			
		} catch (Exception e) {
			Simplex.ExceptionSubject.onNext(e);
			Simplex.Logger.write(e);
		}
		return disposable;
	}
	
	private <TAction extends IAction<TParam, TResult>, TParam, TResult, TActionBinder extends IActionBinder<?, TParam, TResult>> Disposable subscribeInner(
			final TAction action,
			Consumer<TResult> onNext,
			boolean observerOnMainThread, 
			Function<Observable<TResult>,  Observable<TResult>> observable,
			Function<TAction, IChannel> channel,
			boolean preventClone) {
		try {
			
			createChannelIfNull(action);		
						
			Subject<ChannelWrapper<TResult>> wrapperSource = getOrAddObservableSource(action);
			Observable<TResult> source = wrapperSource
				.filter(w -> {
					IChannel ch = getChannel(action, channel);
					for(IChannel targetChannel : ChannelUtil.extract(ch)) {
						for(IChannel sourceChannel : ChannelUtil.extract(w.getChannel())) {
							if (targetChannel.getId().equals(sourceChannel.getId())) {
								return true;
							}
						}
					}
					return false;
				})
				.flatMap(mapper -> Observable.just(mapper.getResult()));
			
			Observable<TResult> o = source;
			if (observable != null) {
				o = observable.apply(source);
			}
			
			if (observerOnMainThread) {
				o = o.observeOn(Simplex.MainThreadScheduler);
			}
			
			return o
					.map(x -> {
						TResult value = x;
						try
				        {
				            //프리미티브, 열거형 타입이거나, 복사 방지 옵션이 설정되었으면 복사하지 않음
							//레퍼런스 타입의 경우 데이터 사본을 전달
				            if (!(preventClone 
				            		|| value.getClass() == Unit.class
				            		|| value.getClass().isPrimitive()
				            		|| Collection.class.isAssignableFrom(value.getClass()))) {
				            	TResult cloned = CloneUtil.deepClone(value);
				                value = cloned;
				            }
				        }
				        catch (Exception ex)
				        {
				            Simplex.ExceptionSubject.onNext(ex);
				            Simplex.Logger.write(ex);
				        }
						return x;
					})
					.subscribe(onNext);
						
		} catch (Exception e) {
			Simplex.ExceptionSubject.onNext(e);
			Simplex.Logger.write(e);
		}
		
		return null;
	}
	
	
	private <TAction extends IAction<TParam, TResult>, TParam, TResult, TActionBinder extends IActionBinder<?, TParam, TResult>> void dispatchInner(
			final TAction action,
			final TParam parameter,
			Action begin,
			Action end, 
			Function<TAction, IChannel> channel) {
		if (begin != null) {
			Simplex.MainThreadScheduler.scheduleDirect(() -> {
				try {
					begin.run();
				} catch (Exception e) {
					Simplex.Logger.write(e);
				}
			});
		}
			
		Action doCompleted = () -> Simplex.MainThreadScheduler.scheduleDirect(() -> {
			if (end != null) {
				try {
					end.run();
				} catch (Exception e1) {
					Simplex.Logger.write(e1);
				}
			}
		});
		
		try {
			
			Observable<TResult> resultObservable = null;
			resultObservable = ((IAction<TParam, TResult>)action).process(parameter);
			
			resultObservable = resultObservable
					.timeout(Simplex.DefaultActionTimeoutMilliseconds, TimeUnit.MILLISECONDS)
					.onErrorResumeNext(throwable -> {
						TransformedResult<TResult> transform = ((IAction<TParam, TResult>)action).transform(throwable);
						if (transform.isTransformed()) {
							Simplex.Logger.write("예외가 발생되지 않고 데이터로 트랜스폼되어 배출되었습니다.");
							return Observable.just(transform.getResult());
						}
						Simplex.Logger.write(throwable);
                        Simplex.ExceptionSubject.onNext(new Exception(throwable));
						return Observable.empty();
					})
					.doOnComplete(doCompleted);
			
			if (action.getClass().getAnnotation(Unsubscribe.class) != null) {
				//더비
				resultObservable.subscribe();
			} else {
				//채널이 생성되지 않았으면 기본 Channel클래스로 생성
                createChannelIfNull(action);
                
                //액션에 해당하는 source 선택
                Subject<ChannelWrapper<TResult>> observer = this.getOrAddObservableSource(action);
                
                //채널 추가
                Observable<ChannelWrapper<TResult>> ob 
                	= resultObservable.flatMap(x -> {
                		return Observable.just(new ChannelWrapper<TResult>(getChannel(action, channel), x));	
                	});

                ob.subscribe(observer::onNext);
                
			}
		} catch(Exception ex) {
			Simplex.ExceptionSubject.onNext(ex);
			Simplex.Logger.write(ex);
			try {
				doCompleted.run();
			} catch (Exception e) {
				Simplex.Logger.write(ex);
			}
		}
	}
	
	private <TParam, TResult> void createChannelIfNull(IAction<TParam, TResult> action) throws IllegalArgumentException, IllegalAccessException {
		Field[] fields = action.getClass().getDeclaredFields();
		for(Field field : fields) {
			if (field.getType() == IChannel.class
					&& field.get(action) == null) {
				field.setAccessible(true);
				field.set(action, new Channel());
			}
		}
	}
			
	@SuppressWarnings("unchecked")
	private <TParam, TResult> Subject<ChannelWrapper<TResult>> getOrAddObservableSource(IAction<TParam, TResult> action) throws Exception {
		Subject<ChannelWrapper<TResult>> subject = null;
		
		// Default채널의 Id를 키로 등록한다.
        String key = action.getDefaultChannel().getId();
        
        if (observableSources.containsKey(key)) {       	
        	subject = (Subject<ChannelWrapper<TResult>>)observableSources.get(key);
        } else {
        	subject = PublishSubject.create();
        	observableSources.put(key, subject);
        }
        return subject;
	}

	private <TAction extends IAction<TParam, TResult>, TParam, TResult> IChannel getChannel(
			TAction action,
			Function<TAction, IChannel> channel) throws Exception {
		IChannel ch = null;
        if (channel != null)
        	ch = channel.apply(action);
        
        if (ch == null)
        	ch = action.getDefaultChannel();
        
        return ch;
	}
	
	public DisposableStore<TActionBinderSet> toDisposableStore(String subscriberId) {
		return new DisposableStore<>(this, subscriberId);
	}
}

class ChannelWrapper<TResult> {
	private IChannel channel;
	
	private TResult result;

	public IChannel getChannel() {
		return channel;
	}

	public void setChannel(IChannel channel) {
		this.channel = channel;
	}

	public TResult getResult() {
		return result;
	}

	public void setResult(TResult result) {
		this.result = result;
	}
	
	ChannelWrapper(IChannel channel, TResult result) {
		this.channel = channel;
		this.result = result;
	}
}

File: Simplex.kotlin/src/main/kotlin/kr/co/tmon/simplex/store/Store.kt
package kr.co.tmon.simplex.store

import kr.co.tmon.simplex.Simplex
import kr.co.tmon.simplex.actions.*
import kr.co.tmon.simplex.annotations.*
import kr.co.tmon.simplex.channels.*
import kr.co.tmon.simplex.extensions.*
import java.util.concurrent.ConcurrentHashMap
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit
import kotlin.reflect.KMutableProperty
import kotlin.reflect.KProperty1
import kotlin.reflect.full.*

class Store<TActionSet>(factory: () -> TActionSet) : IActionStore<TActionSet> {
	companion object {
        inline operator fun <reified TActionSet : Any> invoke() = Store {
            TActionSet::class.java.newInstance()
        }
    }

    private val actions: TActionSet = factory()
    private val observableSources: ConcurrentHashMap<String, Any> = ConcurrentHashMap()

    override fun <TAction : IAction<TResult>, TResult> dispatch(
        action: Function1<TActionSet, IActionBinder<TAction, TResult>>,
        begin: Action?,
        end: Action?,
        channel: Function1<TAction, IChannel>?) {
        dispatchInner(action, null, channel, begin, end)
	}

    override fun <TAction : IParameterizedAction<TParam, TResult>, TParam, TResult> dispatch(
        action: Function1<TActionSet, IParameterizedActionBinder<TAction, TParam, TResult>>,
        parameters: TParam,
        begin: Action?,
        end: Action?,
        channel: Function1<TAction, IChannel>?) {
        dispatchInner(action, parameters, channel, begin, end)
    }

    override fun <TAction : IAction<TResult>, TResult>subscribe(
        action: Function1<TActionSet, IActionBinder<TAction, TResult>>,
        onNext: (TResult) -> Unit,
        observeOnMainThread: Boolean,
        observable : Function1<Observable<TResult>, Observable<TResult>>?,
        channel: Function1<TAction, IChannel>?,
        preventClone: Boolean
    ) : Disposable? {
        return subscribeInner(action, onNext, observable, channel, observeOnMainThread)
    }

    private fun <TAction : IAction<TResult>, TResult> subscribeInner(
        action: Function1<TActionSet, IActionBinder<TAction, TResult>>,
        onNext: (TResult) -> Unit,
        observable: Function1<Observable<TResult>, Observable<TResult>>? = null,
        channel: Function1<TAction, IChannel>? = null,
        observeOnMainThread: Boolean = false) : Disposable {

        val binder = action.invoke(actions)
        //채널이 생성되지 않았으면 기본 Channel클래스로 생성
        createChannelIfNull(binder.action)
        //채널 선택
        val ch = channel?.invoke(binder.action) ?: getDefaultChannel(binder.action)
        //액션에 해당하는 source 선택
        val source = this.getOrAddObservableSource(binder.action)
            .filter { it.first.contains(ch) }
            .unwrap()

        var o = observable?.invoke(source) ?: source
        if (observeOnMainThread)
            o = o.observeOn(Simplex.MainThreadScheduler)

        return o.subscribe{ onNext(it) }
    }

    private fun <TAction : IAction<TResult>, TParam, TResult> dispatchInner(
        action: Function1<TActionSet, IActionBinder<TAction, TResult>>,
        param: TParam,
        channel: Function1<TAction, IChannel>?  = null,
        begin: Action? = null,
        end: Action? = null) {

        Simplex.MainThreadScheduler.run { begin?.run() }
        val doCompleted = Action { Simplex.MainThreadScheduler.run { end?.run() } }

        try {
            val binder = action(actions)
            val resultObservable = binder.action.toObservable(param)
                .timeout(Simplex.DefaultActionTimeout, TimeUnit.MILLISECONDS)
                .onErrorResumeNext { exception: Throwable ->
                    val transform = binder.action.transform(exception)
                    if (transform.isTransformed) {
                        Simplex.Logger?.write?.invoke("예외가 발생되지 않고 데이터로 트랜스폼되어 배출되었습니다.")
                        Observable.just(transform.result)
                    }
                    else {
                        Simplex.Logger?.write?.invoke(exception.message ?: "")
                        Simplex.ExceptionSubject?.onNext(exception)
                        Observable.empty()
                    }
                }
                .doFinally(doCompleted)

            if (binder.action::class.findAnnotation<Unsubscribe>() != null) {
                    createChannelIfNull(binder.action)
                    //더미
                    resultObservable.subscribe()
            }
            else {
                //채널이 생성되지 않았으면 기본 Channel클래스로 생성
                createChannelIfNull(binder.action)
                //채널 선택
                val ch = channel?.invoke(binder.action) ?: getDefaultChannel(binder.action)
                //액션에 해당하는 source 선택
                val observer = this.getOrAddObservableSource(binder.action)
                //채널 추가
                resultObservable.wrap(ch)
                    .subscribe { observer.onNext(it) }
            }
        }
        catch (ex: Exception) {
            Simplex.ExceptionSubject?.onNext(ex)
            Simplex.Logger?.write?.invoke(ex.stackTrace.toString())
            doCompleted.run()
        }
    }


    fun toDisposableStore(subscriberId: String): DisposableStore<TActionSet> {
        return DisposableStore(this, subscriberId)
    }

    private fun <TResult> getDefaultChannel(action: IAction<TResult>): IChannel {
        val abstractAction = action as? AbstractAction<TResult>
        return abstractAction?.default ?: throw IllegalAccessException("${action.javaClass.name}에서 기본 채널을 가져올 수 있는 방법이 구현되지 않았습니다.")
    }

    private fun <TResult> getOrAddObservableSource(action: IAction<TResult>) : PublishSubject<Pair<IChannel, TResult>> {
        // Default채널의 Id를 키로 등록한다.
        val key = getDefaultChannel(action).id

        if (!observableSources.containsKey(key)) {
            val subject:  PublishSubject<Any> = PublishSubject.create()
            observableSources.put(key, subject)
        }

        @Suppress("UNCHECKED_CAST")
        return observableSources.get(key) as PublishSubject<Pair<IChannel, TResult>>
    }

    private fun <TResult> createChannelIfNull(action: IAction<TResult>) {
        for (property in action::class.declaredMemberProperties) {
            @Suppress("UNCHECKED_CAST")
            val channelProperty = property as? KProperty1<IAction<TResult>, IChannel>
            if (channelProperty?.get(action) == null) {
                // 필드 생성
                if (property is KMutableProperty<*>)
                    property.setter.call(action, Channel())
                else throw IllegalAccessException("${action.javaClass.name}액션의 채널 ${property.name}의 생성자에 접근 할 수 없습니다. " +
                            "생성자를 추가하거나 객체를 생성하세요.")
            }
        }
    }

}